src/context/WalletProvider/WalletProvider.tsx

```diff
@@ -515,6 +531,73 @@ export const WalletProvider = ({ children }: { children: React.ReactNode }): JSX
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [state.adapters, state.keyring])
 
+  const resetState = useCallback(() => dispatch({ type: WalletActions.RESET_STATE }), [])
+
+  // Register a MetaMask-like (EIP-1193) provider on wallet connect or load
+  const onProviderChange = useCallback(
+    async (localWalletType: KeyManager | null) => {
+      if (!localWalletType) return
+      setWalletType(localWalletType)
+      if (!walletType) return
+      try {
+        let maybeProvider: InitialState['provider'] = null
+
+        if ([KeyManager.MetaMask, KeyManager.TallyHo].includes(walletType)) {
+          maybeProvider = (await detectEthereumProvider()) as MetaMaskLikeProvider
+        }
+
+        if (walletType === KeyManager.XDefi) {
+          try {
+            maybeProvider = globalThis?.xfi?.ethereum as unknown as MetaMaskLikeProvider
+            state.provider?.on?.('accountsChanged', resetState)
+            state.provider?.on?.('chainChanged', resetState)
+          } catch (error) {
+            throw new Error('walletProvider.xdefi.errors.connectFailure')
+          }
+        }
+        if (walletType === KeyManager.WalletConnect) {
+          const config: WalletConnectProviderConfig = {
+            /** List of RPC URLs indexed by chain ID */
+            rpc: {
+              1: getConfig().REACT_APP_ETHEREUM_NODE_URL,
+            },
+          }
+          maybeProvider = new WalletConnectProvider(config)
+        }
```

Comment:
Not blocking, though I'm not a fan of mixing computations with actions in this way.

A more functional version would have `maybeProvider` as a constant equal to the evaluation of an IIFE, and the provider events mutated by an action function, say `setProviderEvents()'.

Reply:
Great idea! Will add as a stacked PR to keep these lines as a "shuffled things over" diff 🙏🏽
You're right this is one of the cases where a side effects model doesn't make sense, and we have all we need to make it functional

Reply:
https://github.com/shapeshift/web/pull/2593